File: n-reinas.py
# TRABAJO PRÁCTICO FINAL: "Las n-reinas"
# Observaciones:
#   -) Representación del tablero y las posiciones de las reinas:
#       Las posiciones de cada reina son marcadas en una lista L donde cada índice 'i'
#       representa el número de una columna, y el elemento en la posición L[i],
#       representa el número de una fila. Entonces, la posición (fila, columna) de una
#       reina será (L[i], i). Por ejemplo:
#       L=[0,2,3,1] --> Filas
#          0 1 2 3 -->Índices/Número de columna
#       En este ejemplo, la cantidad N de reinas es N=len(L)=4, y éstas se ubican en las
#       posiciones: (0,0), (1,2), (3,2) y (1,3).
#   -) El programa recibe como entrada un número N, que indica la cantidad de reinas que
#      se quieran ubicar en el tablero, y genera como salida un archivo indicando las
#      cantidad de reinas en la primera línea, y en las siguientes N líneas la ubicación
#      'fila columna' de cada reina. En caso de que no haya solución, dicho archivo no es
#       creado y se imprime un mensaje por pantalla indicando que no se pudo hallar la solución.

# vAbs: Int Int --> Int
# vAbs: dados dos enteros x e y devuelve el valor absoluto de x-y.

def vAbs(x,y):
    if x>y:
        return x-y
    else:
        return y-x

# Testing de la función 'vAbs',

def test_vAbs():
    assert vAbs(1,2)==1
    assert vAbs(0,0)==0
    assert vAbs(5,2)==3


# verificar: List Int --> Boolean
# verificar: dada una lista de posiciones y un entero que representa el número de una columna k,
# devuelve True si el tablero con las reinas que están ubicadas hasta la columna k (inclusive),
# con k<longListaDePosiciones, es válido. Caso contrario, devuelve False. Debido a la representación
# elegida para el tablero, no es necesario verificar que haya dos o más reinas en la misma columna, por lo
# tanto, sólo resta verificar que no haya dos o más reinas en la misma fila ni en la misma diagonal.

def verificar(arr, columna):
    verificado=True
    i=0
    while i<=columna and verificado:
        j=i+1
        while j<=columna and verificado:
            if arr[i]==arr[j] or vAbs(arr[i],arr[j])==vAbs(i,j):
                verificado=False
            j+=1
        i+=1
    return verificado

# Testing para la función 'verificar',

def test_verificar():
    assert verificar([0,3,1,2], 2)==True
    assert verificar([0,3,1,2], 3)==False
    assert verificar([0,3,2,2], 1)==True
    assert verificar([0,3,2,2], 2)==False


# reinas: List Int Int --> Boolean
# reinas: dada una lista de posiciones, un número de columna y una cantidad n de reinas,
# ubica, si es posible, las n reinas en un tablero de ajedrez de modo que ninguna se amenace.
# El algoritmo comienza con todos los números de fila de la lista de posiciones en -1. A partir
# de allí a la fila listaDePosiciones[columna] le sumamos 1, y comprobamos (mediante la funcion verificar) si ese numero de fila
# en esa columna lleva a una solución, llamando recursivamente a la funcion reinas(listaPosiciones, columna+1, n).
# Se repite el mismo proceso para columna+1 y en caso de que no resulte en una solución, volvemos a listaPosiciones[columna],
# le asignamos otra fila y vemos si así podemos llegar a una solución.
# Una vez que llegamos a una solución, retornamos True y en la listaPosiciones tendremos los números de fila donde deben
# ubicarse las reinas. En caso de que sea imposible hallarla devolvemos False y la listaPosiciones tendra todos sus elementos
# iguales a -1.

def reinas(arr, col, n):
    ok=False
    if col==n:
        return True
    else:
        while arr[col]<n-1 and not ok:
            arr[col]+=1 # 1) Le asignamos a la reina un número de fila.
            if verificar(arr,col): # Si la posición es validada llamamos recursivamente a la función tratando de colocar la siguiente reina y vemos si nos lleva a una solución.
                ok=reinas(arr,col+1,n)   # Si ok es False ese número de fila no lleva a una solución y debemos volver al paso 1).
        if ok==False:
            arr[col]=-1
        return ok

# Testing para la función 'reinas'.

def test_reinas():
    assert reinas([-1], 0, 1)==True
    assert reinas([-1,-1], 0, 2)==False
    assert reinas([-1,-1,-1], 0, 3)==False
    assert reinas([-1,-1,-1,-1,-1], 0, 5)==True


# nreinas: None --> None
# nreinas: Recibe un número n ingresado por el usuario, llama a la funcion 'reinas'
# que se encarga de crear la solución al problema de ubicar las n reinas en un
# tablero de ajedrez sin que se ataquen. Una vez hallada la solución, si es que esta existe,
# se escribe en un archivo con la cantidad de reinas y las posiciones 'fila columna' de cada
# una de ellas. Si no existe solución, el archivo no se crea y se le informa al usuario por
# pantalla que no ha sido posible hallar una solución.

def nreinas():
    n=int(input("Ingrese la cantidad n de reinas: "))
    solucion=[-1]*n
    sol=reinas(solucion,0,n)
    if sol==False:
        print("No se pudo encontrar solución.")
    else:
        # Escribimos la solución en un archivo preparado para ser usado como entrada en el programa de C.
        solReinas=open("solucionReinas.txt", "w")
        solReinas.write(str(n)+'\n') #Cantidad de reinas.
        for i in range(0, n): # Escribimos en el archivo la fila y columna de cada reina separadas por un espacio.
            solReinas.write(str(solucion[i])+' '+str(i)+'\n')
        print("¡Solución guardada en 'solucionReinas.txt'!")

nreinas()

File: n-reinas.c
#include <stdio.h>
#include <math.h>

/* TRABAJO PRÁCTICO FINAL: "Las n-reinas"

Observaciones:

1) Ingreso de la cantidad y posiciones de las reinas:
    la cantidad N de reinas y las posiciones de las mismas serán leídas desde un archivo '.txt'.
    en la primera línea del archivo se colocará la cantidad N de reinas, y en las siguientes N
    líneas la fila y columna de cada reina separadas por UN solo espacio. Por ejemplo:

               4   --> Cantidad N de reinas.
      fila <-- 0 0 --> columna
               2 1
               3 3
               1 2

    La entrada anterior nos índica en la primera línea que hay 4 reinas y en las siguientes
    4 líneas está la posición 'fila columna' de cada reina, (primero se escribe la fila y separado
    de un espacio, la columna). De todos modos, hay un archivo adjunto 'ejemploPosiciones' que sirve como 
    ejemplo para ver la forma de indicar la cantidad y posiciones de las reinas.


2) Representación del tablero y de las posiciones:
    Las posiciones de las reinas en el tablero son representadas en una matriz bidimensional 
    de tamaño Nx2. Sea i<N, entonces el elemento de la matriz[i][0], representa el número
    de fila de la reina 'i' y el elemento matriz[i][1] representa el número de columna
    de la reina 'i'. Por ejemplo:

    m[4][2]={{0,0},{2,1},{3,3},{1,2}}

    Siguiendo el ejemplo mencionado en el punto 1) el array 'm' nos dice que tenemos 4 reinas
    y éstas se ubican en las posiciones: ({fila, columna}) {0,0}, {2,1}, {3,3} y {1,2}.

3) Una vez leídas las posiciones desde el archivo, verificamos el tablero con la función 'verificaTablero',
   y mostramos por pantalla un mensaje indicando si la solución propuesta es válida o no.


*/


// verificaTablero: Array Int --> Int
// verificaTablero: recibe un array con las posiciones de las n-reinas en el tablero
// y la cantidad de reinas, devuelve 1 si las reinas no se amenazan entre ellas 
// (esto es si ninguna reina está en la misma fila, columna o diagonal).
// En caso contrario devuelve 0 y muestra por pantalla un mensaje indicando
// cuáles son las reinas que se están amenazando entre sí.

int verificaTablero(int listaPosiciones[17][2], int cantReinas){
    int i=0;
    int verificado=1;
    while(i<cantReinas && verificado==1){
        int j;
        j=i+1;
        while(j<cantReinas && verificado==1){
            if(listaPosiciones[i][0]==listaPosiciones[j][0] || //Verifica si hay dos filas iguales.
                listaPosiciones[i][1]==listaPosiciones[j][1] ||  //Verifica si hay dos columnas iguales.
                fabs(listaPosiciones[i][0]-listaPosiciones[j][0])==fabs(listaPosiciones[i][1]-listaPosiciones[j][1])){ //Verifica que no estén en la misma diagonal

                verificado=0;
                if(listaPosiciones[i][0]==listaPosiciones[j][0]){ // Hay dos reinas en la misma fila.
                    printf("El tablero no es correcto, las reinas (%d,%d) y (%d,%d), se encuentran en la misma fila.\n", listaPosiciones[i][0],listaPosiciones[i][1],listaPosiciones[j][0],listaPosiciones[j][1]);
                }
                if(listaPosiciones[i][1]==listaPosiciones[j][1]){ //Dos reinas en la misma columna
                    printf("El tablero no es correcto, las reinas (%d,%d) y (%d,%d), se encuentran en la misma columna.\n", listaPosiciones[i][0],listaPosiciones[i][1],listaPosiciones[j][0],listaPosiciones[j][1]);
                }
                if(fabs(listaPosiciones[i][0]-listaPosiciones[j][0])==fabs(listaPosiciones[i][1]-listaPosiciones[j][1])){ // Hay dos reinas en la misma diagonal
                    printf("El tablero no es correcto, las reinas (%d,%d) y (%d,%d), se encuentran en la misma diagonal.\n", listaPosiciones[i][0],listaPosiciones[i][1],listaPosiciones[j][0],listaPosiciones[j][1]);
                }
            }
            j++;
        }
        i++;
    }
    if(verificado==1){
      printf("¡La solución es válida! :)\n");
    }
    return verificado;
}


// En la función 'main' pedimos el ingreso del nombre del archivo, leemos los datos y almacenamos las ubicaciones de las reinas
// en el array 'listaPosiciones'. Una vez cerrado el archivo, llamamos a la función que verificará las 
// posiciones almacenadas en el array y devolverá un mensaje por pantalla indicando si la solución es correcta
// o no (en caso de no serlo, se detalla el por qué).

int main(){
    char nombreArchivo[30];
    printf("Ingrese el nombre del archivo con '.txt':\n");
    scanf("%s", nombreArchivo);

    // Abrimos el archivo.
    FILE *archivo;
    archivo=fopen(nombreArchivo, "r");
    // En la primera línea del archivo leemos la cantidad n de reinas.
    int cantReinas;
    fscanf(archivo, "%d", &cantReinas);
    int listaPosiciones[cantReinas][2];
    // En las restantes n lineas del archivo leemos las posiciones 'fila columna' de las reinas
    // y las agregamos al array de enteros.
    int linea=0;
    int error=0;
    while (linea<=cantReinas && error==0){ //Si alguna fila o columna está fuera de rango mostramos el error en pantalla.
        int fila, columna;
        fscanf(archivo, "%d %d", &fila, &columna);
        if(fila>=cantReinas || fila<0){
            printf("Error en el archivo: el número de fila: %d es incorrecto.\n", fila);
            error=1;
        }
        if(columna>=cantReinas || columna<0){
            printf("Error en el archivo: el número de columna: %d es incorrecto.\n", columna);
            error=1;
        }
        else{
            listaPosiciones[linea][0]=fila;
            listaPosiciones[linea][1]=columna;
            linea++;
        }
        
    }
    fclose(archivo);

    // Llamamos a la función que realiza la verificación y mostrará el resultado por pantalla 
    // si no hubo error alguno en el ingreso de los números de filas.
    if(error!=1){
        verificaTablero(listaPosiciones, cantReinas);
    }

    return 0;
}
